### PR TITLE
Added Unit Test and Shortened cardinalToDegree() Lookup Function

### PR DESCRIPTION
**Pull request by Beakerboy**
Refactored function. Added tests. Added documentation to function.

### CODE REVIEW

File: test/buildingpart.test.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { BuildingPart } from '../src/buildingpart.js';
+
+test('Test Cardinal to Degree', () => {
+  expect(BuildingPart.cardinalToDegree('N')).toBe(0);
+  expect(BuildingPart.cardinalToDegree('sSw')).toBe(202);
+  expect(BuildingPart.cardinalToDegree('Wse')).toBeUndefined();
+});
```
Comment (justinormont) [inline test/buildingpart.test.js:7]: Unit test is good; there are couple more cases that could be covered:
```suggestion
import { BuildingPart } from '../src/buildingpart.js';

test('Test Cardinal to Degree', () => {
  expect(BuildingPart.cardinalToDegree('N')).toBe(0);
  expect(BuildingPart.cardinalToDegree('sSw')).toBe(202);
  expect(BuildingPart.cardinalToDegree('Wse')).toBeUndefined();
  expect(BuildingPart.cardinalToDegree(0)).toBeUndefined();
  expect(BuildingPart.cardinalToDegree(null)).toBeUndefined();
  [['N', 0], ['NE', 45], ['E', 90], ['SE', 135], ['S', 180], ['SW', 225], ['W', 270], ['NW', 315], ['NNE', 22], ['ENE', 67], ['ESE', 112],
    ['SSE', 157], ['SSW', 202], ['WSW', 247], ['WNW', 292], ['NNW', 337]].forEach((a) => {
      expect(BuildingPart.cardinalToDegree(a[0])).toBe(a[1]);
    });
});
```

Tossed in an exhaustive test of expected values, plus a couple cases that should also be `undefined`.

File: src/buildingpart.js [hunk 2]
```diff
@@ -407,60 +407,22 @@ class BuildingPart {
...
-    if (cardinalUpperCase === 'NNW') {
-      return 337;
+    const index = 'N NNE NE ENE E ESE SE SSE S SSW SW WSW W WNW NW NNW'.split(' ').indexOf(cardinalUpperCase);
+    if (index === -1) {
+      return undefined;
     }
-    return undefined;
+    const degreesTimesTwo = index * 45;
+    // integer floor
+    return degreesTimesTwo % 2 === 0 ? degreesTimesTwo / 2 : (degreesTimesTwo - 1) / 2;
```
Comment (justinormont) [inline src/buildingpart.js:425]: Code is yours, so it's up to you. My general opinion is the new code is too complex and therefore harder to maintain, and the code is slower than the original.

The other method I was considering was a JS Map:

```js
const cardinalToDegreeLookup = new Map([['N', 0], ['NE', 45], ['E', 90], ['SE', 135], ['S', 180], ['SW', 225], ['W', 270], ['NW', 315], ['NNE', 22], ['ENE', 67], ['ESE', 112], ['SSE', 157], ['SSW', 202], ['WSW', 247], ['WNW', 292], ['NNW', 337]]);

/**
 * Convert a cardinal direction to degrees.
 * North is zero and values increase clockwise.
 *
 * @param {string} cardinal - the direction.
 *
 * @return {int} degrees
 */
static cardinalToDegree(cardinal) {
  const cardinalUpperCase = `${cardinal}`.toUpperCase();
  if (cardinalToDegreeLookup.has(cardinalUpperCase)) { return cardinalToDegreeLookup.get(cardinalUpperCase); }
  return undefined;
}
```

The `Map()` style is clean and understandable, but slower than using a `if/return, if/return, if/return` pattern. This function is only called when a direction is given, so speed likely isn't too important.

Comment (justinormont) [reply src/buildingpart.js:425]: Technically, the `.has()` is unneeded as you could just have `return cardinalToDegreeLookup.get(cardinalUpperCase);` as `Map.get()` returns `undefined` for missing keys.

Generally I prefer to make the `return undefined` explicit.

Comment (Beakerboy) [reply src/buildingpart.js:425]: I felt the in-then-else was just too bulky. I’m a math guy, not a CS guy, so my first instinct was to reduce it to a simple math problem. If you say it’s slow, I trust you. I like the map idea. If you had originally submitted that, I probably wouldn’t have messed with it.

**Pull request closed without merging**